src/predicate/mod.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2017, Nick Stevens <nick@bitcurry.com>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/license/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Predicate
+//!
+//! This module contains the `Predicate` trait and a number of combinators for
+//! the trait. See the crate docs, and the docs for `Predicate`, for full detail.
+
+// primitive `Predicate` types
+pub mod constant;
+pub mod ord;
+pub mod set;
+pub use self::constant::{always, never, BooleanPredicate};
+pub use self::ord::{eq, ne, lt, le, gt, ge, EqPredicate, OrdPredicate};
+pub use self::set::{in_set, SetPredicate};
+
+// combinators
+pub mod boolean;
+pub use self::boolean::{AndPredicate, OrPredicate, NotPredicate};
+
+/// Trait for generically testing a type against a dynamically created
+/// predicate function.
+///
+/// The exact meaning of `test` depends on the situation, but will usually
+/// mean that the tested item is in some sort of pre-defined set.  This is
+/// different from `Ord` and `Eq` in that an `item` will almost never be the
+/// same type as the implementing `Predicate` type.
+pub trait Predicate {
+    /// The type that this `Predicate` will accept for testing.
+    type Item;
+
+    /// Execute this `Predicate` against `variable`, returning the resulting
+    /// boolean.
+    fn test(&self, variable: &Self::Item) -> bool;
```

Comment:
Yeah, I wasn't thrilled with `test` either. But `check` doesn't really fit either, since it implies correctness, and we don't carry any particular notion of correctness. And `apply` has other baggage associated with it. I'm open to suggestions and will continue to think about it as well.